authorization/jwt-validation.md

```diff
@@ -9,24 +9,63 @@ menu:
 {{< terminal >}}
 GET /resource HTTP/1.1
 Host: krakend.example.com
-Authorization: Bearer eyJhbGciOiJIUzI1NiIXVCJ9...TJVA95OrM7E20RMHrHDcEfxjoYZgeFONFh7HgQeyJhbGciOiJIUzI1NiJ9.eyJzdWIiOiIxMjM0NTY3ODkwIiwibmFtZSI6IktyYWtlbkQiLCJpYXQiOjE1MTYyMzkwMjJ9.NVFYj2MhyvJjMESOg4ktIOfzak2ekD7IrCa9-UiO4QA
+Authorization: Bearer eyJhbGciOiJIUzI1NiIXVCJ9.(truncated).ktIOfzak2ekD7IrCa9-UiO4QA
 {{< /terminal >}}
-**And cookies?** Yes, you can also send the validation token in a **cookie**.
 
-We consider a JWT token to be valid when is well formed, signed by a recognized issuer, unexpired, with some claims and is not marked as [revoked](/docs/authorization/revoking-tokens).
+Or optionally, you can send the token **inside a cookie** instead.
+
+### JWT header requirements
+When KrakenD decodes the `base64` token string passed in the `Bearer` or the cookie, it expects to find in its **header** section the following 3 fields:
+
+	{
+	  "alg": "RS256",
+	  "typ": "JWT",
+	  "kid": "MDNGMjU2M0U3RERFQUEwOUUzQUMwQ0NBN0Y1RUY0OEIxNTRDM0IxMw"
+	}
+
+The values of the `alg` and `kid` depend on your implementation, but they must be present.
 
-{{< note title="A note on JWT generation" >}}
-When you generate tokens for end-users, make sure to set a **low expiration**. Tokens are supposed to have short lives and are recommended to expire in minutes or hours.
+The value provided in the `kid` string **must match the `kid` value of the associated URL provided in the `jwk-url`** to verify the signature. The example above used [this public key](https://albert-test.auth0.com/.well-known/jwks.json), notice how the `kid` matches both in the URL and in the token. Also notice that the `kid` is `base64` encoded, so we didn't write `03F2563E7DDEAA09E3AC0CCA7F5EF48B154C3B13` (which is the real Key ID).
+
+KrakenD is built with security in mind and uses JWS (instead of plain JWT or JWE), and the `kid` points to the right signature in the JWS. This is why this entry is mandatory to validate your tokens. 
+
+{{< note title="Important!" >}}
+Make sure you are declaring the right `kid` in your JWT. Paste a token in a [debugger](https://jwt.io/#debugger-io) to find out.
 {{< /note >}}
 
+### Validation process
+KrakenD does the following validation to let users hit protected endpoints:
+
+- The `jwk-url` must be accessible by KrakenD at all times (caching is available)
+- The token is [well formed](https://jwt.io/#debugger-io)
+- The `kid` in the header is listed in the `jwk-url`.
+- The content of the JWK Keys (`kid`) is **base64** urlencoded
+- The algorithm `alg` is supported by KrakenD and matches exactly the one used in the endpoint definition.
+- The token hasn't expired
+- The signature is valid.
+- The given `issuer` matches (if present)
+- The given `audience` matches (if present)
+- The given claims are within the endpoint accepted `roles` (if present)
+
+The configuration allows you to define permissions per endpoint.
```

Comment:
set of roles required. a token with role A and B could access endpoints requiring roles B or C